[Final Project] TorchSplit: a Compiler Analysis for Graph Decomposition and Parallel Execution on Neural Networks

az275 opened the pull request:
Closes #604

az275 committed:
final project blog post

az275 committed:
add links

az275 committed:
minor edits

az275 committed:
formatting

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
```diff
@@ -0,0 +1,76 @@
++++
+title = "Torchsplit: a Compiler Analysis for Graph Decomposition and Parallel Execution on Neural Networks"
```
Elsewhere, it's capitalized "TorchSplit."

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
Is there a link you can include to more information about Torch FX and/or the FX IR? Maybe something like [this](https://github.com/pytorch/pytorch/blob/main/torch/fx/README.md)?

jeffreyqdd replied:
Linked.

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
I don't fully understand the role that phi-nodes play here… from your description so far, this is "just" a data-flow graph, with no control flow and no side effects. Without a CFG, I'm not sure what a phi-node does?

jeffreyqdd replied:
Updated the writeup to explain.

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
This might be a good place to include a teensy bit more detail about how your implementation works. Some questions this raised for me are:

* How did you generate inputs for the components? (I understand there are concrete inputs for the whole program, but I don't see where they come from for individual components in isolation.)
* How did you cap the memory?
* What are the units for "achievable throughput"?
* How many different memory sizes is enough to get a useful function?

jeffreyqdd replied:
Updated the writeup!

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
Some links to Gurobi and gurobipy might be nice.

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
A slightly more detailed sentence would help people know what to click on… maybe something like:

> Our open-source implementation is split across two repositories: one for [performing the split] and one for [actually running the split model].

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
What is a "Perlmutter node"?

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
It seems like a good idea to describe these limitations above, when you're introducing the system (rather than just here in the eval). That would make them a bit easier to understand. For example, as it stands I don't completely know what you mean by "we set the GPU memory requirements for each component to be 8GB"… does this mean that the profiler doesn't actually sweep memory caps as described above?

az275 replied:
Opted to leave this here since it's actually not an inherent limitation of the profiler or the ILP formulation. We made the simplification because of the characteristics of the CLIP model and because we ran out of time to get batching working. Edited for clarity!

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
Any chance you measured how long the splitting analysis took? No big deal if not, but it would be cool to say something about it if yo have the numbers. It could help point the way toward understanding whether this is going to work for more complicated/larger models.

sampsyo commented on content/blog/2025-12-15-torchsplit.md:
What is "MIG partitioning"?

sampsyo reviewed (changes requested):
Great work! This was clearly described, and it's awesome that you got one model working on your framework end-to-end. I'm excited to see whether this approach scales to bigger models that might be harder to partition manually.

I have a few low-level writing suggestions—could you please address those and then use GitHub's "request review" button to let me know when you're ready to publish?

az275 committed:
address some comments

jeffreyqdd committed:
fix: address comments

- elaborate on phi nodes
- addresses how to partition generation details
- add brief section on end-to-end runtime

jeffreyqdd committed:
fix: incorrect link syntax

sampsyo reviewed (approved):
Looking good; nice work again!
